### ISSUE TITLE
Intermediate change scenario

### ISSUE THREAD

**Issue by csosto-pk** [csosto-pk/tls-suppress-intermediates#2]
From https://mailarchive.ietf.org/arch/msg/tls/KAAKiEki36gL8g40ZimNHk39OXY/ 

Suppose the following sequence of events happen:

1: A CA uses a new intermediate for reasons (no longer cross-signing, etc.)
2: A site gets a certificate from the new intermediate.
3: An older firefox version connects and thinks it knows all the
certificates in the world.

This would seem to break and it wasn't clear to me how this would be
handled.

**Comment by csosto-pk** [comment]
From https://mailarchive.ietf.org/arch/msg/tls/sJ4vlchFfKtKqYDADdxwAWoR2ug/ 

I think the client would have to fall back and retry the TLS connection
without requesting that intermediates be omitted. In general, I think this
is the only reliable answer as AIA-chasing doesn't always work. (Either the
AIA server can be down, or the chain can be from a private CA that doesn't
support AIA.)

**Comment by csosto-pk** [comment]
I added text to add an optional solution to this issue
> EDNOTE: One additional option could be to use the TLS extension defined
in [RFC7924] to include the chain fingerprint so the peer can confirm
that he does not need to send the chain because the peer asking for
suppression has the correct chain to validate the server.
That could prevent inadvertent mistakes where the client thinks it has
the intermediates to validate the server, but what it has is wrong.

We should discuss this and [Adam L's suggestion](https://mailarchive.ietf.org/arch/msg/tls/sJ4vlchFfKtKqYDADdxwAWoR2ug/) 
>I think the client would have to fall back and retry the TLS connection
without requesting that intermediates be omitted.

in the WG if the draft get adopted

**Comment by csosto-pk** [comment]
In https://mailarchive.ietf.org/arch/msg/tls/Zp4GfjXaCrtmVwfVEst-iD_-Njw/ Ryan S. was pointing out that if we are willing to assume that there is a mechanism for the ICA list to be delivered OOB then maybe we can simplify the failure scenario.

**Comment by csosto-pk** [comment]
Related issues 
- https://github.com/csosto-pk/tls-suppress-intermediates/issues/6 talks about constrained intermediates and how they are not available and cause failures. What assumptions we could make and potential solution to have the server send the chain regardless if it knows it has a constrained intermediate. 
- https://github.com/csosto-pk/tls-suppress-intermediates/issues/4 talks about a potential solution for the server to keep state in a ticket so it knows the client has the ICA or not. Not sure at this time if it is worth the effort.

**Comment by csosto-pk** [comment]
As Chris Patton pointed out to Bas, there is precedent for the abort-and-retry: For ECH when the server can't decrypt the inner CH, it aborts the connection and the client tries again.

**Comment by csosto-pk** [comment]
David Benjamin also made a good point in IETF-113 which is that because of the TBD3 freshness requirement practically we are taking WebPKI out because most clients are not up to date, so they will not be fresh and thus not use the feature. That had not occurred to me. 

So, I am thinking if we want this to work in PKI we probably cannot require the ICA to be fresh. To make it work for the Web we 
- either have to have a way to avoid the false start as we had discussed before 
- or push for an external ICA list source of truth (e.g. CCADB) which could be used by the sender to confirm if its ICA are in the list
- have a fallback mechanism like what you are describing in ECH

**Comment by martinthomson** [comment]
If clients only need to have an update within the last week, some/many might be able to use the feature.

**Comment by csosto-pk** [comment]
Clarification on David B.'s comment 

> So the idea is that you rely on the client to shut off the optimization after a period of time when stale? That's less immediately broken, but still seems questionable. It's needlessly flaky for clients that don't manage to update their list. Clients that don't get updates either can't participate in this mechanism or have to drop out immediately.
> 
> It also gets complicated with user-added roots. A client may have intermediates for publicly-trusted roots preloaded, but probably doesn't have them for user-added roots. Now, you could say, oh, we assume the server knows whether it's using one the client believes is preloaded. But that makes our already rather flaky trust anchor agility story even more flaky. Publicly-accessible servers, after all, have to support more than one client.
> 
> The pre-IETF version of QUIC had a version of this mechanism, but it was much more robust. It gave names to preloaded intermediate sets. While more work to specify, I think that's a better and more robust direction to go here.

**Comment by csosto-pk** [comment]
> Clarification on David B.'s comment
> 
> > So the idea is that you rely on the client to shut off the optimization after a period of time when stale? 

Yes, but that could lead to inadvertent failures. Ryan Sleevi was pointing out that these are problematic, so we were also thinking that if there was a third party like CCADB to host these ICA lists then the server could also make a decision to send its ICAs regardless if it knows they don’t exist in the list. That should prevent even more failures. 

> > It's needlessly flaky for clients that don't manage to update their list. Clients that don't get updates either can't participate in this mechanism or have to drop out immediately.

Right. It is a tradeoff for picking the freshness requirement TBD3-time the prevents failures but also gets to prevent sending ICAs as much as possible. We plan to study the WebPKI ICA list change frequency for the ICA list to get more concrete data before thinking about the best TBD3. 

> > It also gets complicated with user-added roots. A client may have intermediates for publicly-trusted roots preloaded, but probably doesn't have them for user-added roots. 

True, isn’t that usecase kind of rare though? 
If we manage to have a common CCADB ICA list (for WebPKI) then the server could use that to make the decision itself, but I get your point. 

Now, you could say, oh, we assume the server knows whether it's using one the client believes is preloaded. But that makes our already rather flaky trust anchor agility story even more flaky. Publicly-accessible servers, after all, have to support more than one client.

> > The pre-IETF version of QUIC had a version of this mechanism, but it was much more robust. It gave names to preloaded intermediate sets. While more work to specify, I think that's a better and more robust direction to go here.

Agreed. We probably will investigate that with CCADB. Probably we should keep it simple if possible without defining list versions, but at the same time we want as much use as possible. If CCADB is receptive to it, I think we can come up with something.

**Comment by csosto-pk** [comment]
Mechanism that Bas suggested 

> A browser will only send the flag if its list of intermediates has been updated within the last 24 hours.  When a server receives the flag, it will suppress the intermediates unless all the included SCTs for the leaf are less than 96 hours old. That leaves the browser vendor 24 hours to detect a missing intermediate and/or ship it into its list. An obvious downside is that we're serving the intermediates the first 96 hours, but that can be prevented with reissue by having certbot pre-request the certificate. There are other reasons that certbot might want to do this — for instance to include the proof-of-inclusion into the certificate instead of a bigger post-quantum SCT.
